test/others/minimax.spec.js

```diff
@@ -0,0 +1,170 @@
+
+    while (!game.isGameOver(state)) {
+      if (state.turn === 'x') {
+        state = game.nextState(state, minimaxFor(state.turn, state).move);
+      } else {
+        const move = game.emptyCells(state)[0];
+        state = game.nextState(state, move);
+      }
+    }
+
+    expect(game.isGameOver(state)).toBe(true);
+    expect(game.getScore(state)).toEqual({x: 1, o: 0});
+  });
+
+  it('should win versus dumb agent as second player', function () {
+    let state = game.newState('x');
+
+    while (!game.isGameOver(state)) {
+      if (state.turn === 'o') {
+        state = game.nextState(state, minimaxFor(state.turn, state).move);
+      } else {
+        const move = game.emptyCells(state)[0];
+        state = game.nextState(state, move);
+      }
+    }
+
+    expect(game.isGameOver(state)).toBe(true);
+    expect(game.getScore(state)).toEqual({x: 0, o: 1});
+  });
+
+
+  it('should be a tie for two minimax agents', function () {
+    let state = game.newState('x');
+
+    while (!game.isGameOver(state)) {
+      state = game.nextState(state, minimaxFor(state.turn, state).move);
+    }
+    expect(game.isGameOver(state)).toBe(true);
+    expect(game.getScore(state)).toEqual({x: 0, o: 0});
+  });
+});
+
+function ticTacToe() {
+  'use strict';
+
+  function newState(turn) {
+    return {
+      board: [[0, 0, 0],
+        [0, 0, 0],
+        [0, 0, 0]],
+      turn
+    };
+  }
+
+  function emptyCells(state) {
+    const result = [];
+    state.board.forEach((row, y) => {
+      row.forEach((cell, x) => {
+        if (cell === 0) {
+          result.push({x, y})
+        }
+      });
+    });
+
+    return result;
+  }
+
+  function getWinner(state) {
+    const winVariants = [
+      [{x: 0, y: 0}, {x: 0, y: 1}, {x: 0, y: 2}],
+      [{x: 1, y: 0}, {x: 1, y: 1}, {x: 1, y: 2}],
+      [{x: 2, y: 0}, {x: 2, y: 1}, {x: 2, y: 2}],
+
+      [{x: 0, y: 0}, {x: 1, y: 0}, {x: 2, y: 0}],
+      [{x: 0, y: 1}, {x: 1, y: 1}, {x: 2, y: 1}],
+      [{x: 0, y: 2}, {x: 1, y: 0}, {x: 2, y: 2}],
+
+      [{x: 0, y: 0}, {x: 1, y: 1}, {x: 2, y: 2}],
+      [{x: 2, y: 0}, {x: 1, y: 1}, {x: 2, y: 0}],
+    ];
+
+    for (const variant of winVariants) {
+      const combo = variant.map(cell => state.board[cell.y][cell.x]).join('');
+      if (combo === 'xxx') {
+        return 'x';
+      } else if (combo === 'ooo') {
+        return 'o';
+      }
+    }
+
+    return null;
+  }
+
+  function allFieldsMarked(state) {
+    return state.board.every(row => row.every(cell => cell !== 0));
+  }
+
+  function isGameOver(state) {
+    return allFieldsMarked(state) || getWinner(state) !== null;
+  }
+
+  function getScore(state) {
+    if (getWinner(state) === 'x') {
+      return {x: 1, o: 0};
+    } else if (getWinner(state) === 'o') {
+      return {x: 0, o: 1};
+    }
+
+    return {x: 0, o: 0};
+  }
+
+  function nextState(state, move) {
+    const board = state.board;
+    return {
+      board: [
+        ...board.slice(0, move.y),
```

Comment:
This is a bit hard to read. Something like this for updating the board should be easier to understand:

```ts
const newBoard = JSON.parse(JSON.stringify(state.board));
newBoard[move.x][move.y] = state.turn;
```

Reply:
Clever! Fixed.